// component-playground/lib/components/doc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypesArray = [{
  key: "array",
  test: _react.PropTypes.array,
  isRequired: _react.PropTypes.array.isRequired
}, {
  key: "boolean",
  test: _react.PropTypes.bool,
  isRequired: _react.PropTypes.bool.isRequired
}, {
  key: "function",
  test: _react.PropTypes.func,
  isRequired: _react.PropTypes.func.isRequired
}, {
  key: "number",
  test: _react.PropTypes.number,
  isRequired: _react.PropTypes.number.isRequired
}, {
  key: "object",
  test: _react.PropTypes.object,
  isRequired: _react.PropTypes.array.isRequired
}, {
  key: "string",
  test: _react.PropTypes.string,
  isRequired: _react.PropTypes.string.isRequired
}, {
  key: "node",
  test: _react.PropTypes.node,
  isRequired: _react.PropTypes.node.isRequired
}, {
  key: "element",
  test: _react.PropTypes.element,
  isRequired: _react.PropTypes.element.isRequired
}];

var getReactPropType = function getReactPropType(propTypeFunc) {
  var name = "custom";
  var isRequired = false;

  propTypesArray.some(function (propType) {
    if (propTypeFunc === propType.test) {
      name = propType.key;
      return true;
    }
    if (propTypeFunc === propType.isRequired) {
      name = propType.key;
      isRequired = true;
      return true;
    }
    return false;
  });
  return { name: name, isRequired: isRequired };
};

var Doc = function (_Component) {
  _inherits(Doc, _Component);

  function Doc() {
    _classCallCheck(this, Doc);

    return _possibleConstructorReturn(this, (Doc.__proto__ || Object.getPrototypeOf(Doc)).apply(this, arguments));
  }

  _createClass(Doc, [{
    key: "render",
    value: function render() {

      var propTypes = [];
      var _props = this.props,
          componentClass = _props.componentClass,
          ignore = _props.ignore,
          propDescriptionMap = _props.propDescriptionMap;

      for (var propName in componentClass.propTypes) {
        if (ignore.indexOf(propName)) {
          propTypes.push({
            propName: propName,
            type: getReactPropType(componentClass.propTypes[propName]),
            description: propDescriptionMap[propName] || ""
          });
        }
      }

      return _react2.default.createElement(
        "div",
        null,
        _react2.default.createElement(
          "ul",
          null,
          propTypes.map(function (propObj) {
            return _react2.default.createElement(
              "li",
              { key: propObj.propName },
              _react2.default.createElement(
                "b",
                null,
                propObj.propName
              ),
              _react2.default.createElement(
                "i",
                null,
                ": " + propObj.type.name
              ),
              propObj.description && " - " + propObj.description,
              _react2.default.createElement(
                "b",
                null,
                "" + (propObj.type.isRequired ? " required" : "")
              )
            );
          })
        )
      );
    }
  }]);

  return Doc;
}(_react.Component);

Doc.defaultProps = {
  propDescriptionMap: {},
  ignore: []
};
Doc.propTypes = {
  componentClass: _react.PropTypes.func,
  propDescriptionMap: _react.PropTypes.object,
  ignore: _react.PropTypes.array
};
exports.default = Doc;
