// component-playground/lib/components/es6-preview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; /* eslint new-cap:0 no-unused-vars:0 */


var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _reactDom = require("react-dom");

var _babelStandalone = require("babel-standalone");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var getType = function getType(el) {
  var t = typeof el === "undefined" ? "undefined" : _typeof(el);

  if (Array.isArray(el)) {
    t = "array";
  } else if (el === null) {
    t = "null";
  }

  return t;
};

var wrapMap = {
  wrapnumber: function wrapnumber(num) {
    return _react2.default.createElement(
      "span",
      { style: { color: "#6170d5" } },
      num
    );
  },
  wrapstring: function wrapstring(str) {
    return _react2.default.createElement(
      "span",
      { style: { color: "#F2777A" } },
      "'" + str + "'"
    );
  },
  wrapboolean: function wrapboolean(bool) {
    return _react2.default.createElement(
      "span",
      { style: { color: "#48A1CF" } },
      bool ? "true" : "false"
    );
  },
  wraparray: function wraparray(arr) {
    return _react2.default.createElement(
      "span",
      null,
      "[",
      arr.map(function (entry, i) {
        return _react2.default.createElement(
          "span",
          { key: i },
          wrapMap["wrap" + getType(entry)](entry),
          i !== arr.length - 1 ? ", " : ""
        );
      }),
      "]"
    );
  },
  wrapobject: function wrapobject(obj) {
    var pairs = [];
    var first = true;

    for (var key in obj) {
      pairs.push(_react2.default.createElement(
        "span",
        { key: key },
        _react2.default.createElement(
          "span",
          { style: { color: "#8A6BA1" } },
          (first ? "" : ", ") + key
        ),
        ": ",
        wrapMap["wrap" + getType(obj[key])](obj[key])
      ));

      first = false;
    }

    return _react2.default.createElement(
      "i",
      null,
      "Object {",
      pairs,
      "}"
    );
  },
  wrapfunction: function wrapfunction() {
    return _react2.default.createElement(
      "i",
      { style: { color: "#48A1CF" } },
      "function"
    );
  },
  wrapnull: function wrapnull() {
    return _react2.default.createElement(
      "span",
      { style: { color: "#777" } },
      "null"
    );
  },
  wrapundefined: function wrapundefined() {
    return _react2.default.createElement(
      "span",
      { style: { color: "#777" } },
      "undefined"
    );
  }
};

var EsPreview = function (_Component) {
  _inherits(EsPreview, _Component);

  function EsPreview() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, EsPreview);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = EsPreview.__proto__ || Object.getPrototypeOf(EsPreview)).call.apply(_ref, [this].concat(args))), _this), _initialiseProps.call(_this), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(EsPreview, [{
    key: "render",
    value: function render() {
      return _react2.default.createElement("div", { ref: "mount" });
    }
  }]);

  return EsPreview;
}(_react.Component);

EsPreview.propTypes = {
  code: _react.PropTypes.string.isRequired,
  scope: _react.PropTypes.object.isRequired
};

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this._compileCode = function () {
    var _props = _this2.props,
        code = _props.code,
        scope = _props.scope;

    return (0, _babelStandalone.transform)("\n      ((" + Object.keys(scope).join(",") + ") => {\n        var list = [];\n        var console = { log(...x) {\n          list.push({val: x, multipleArgs: x.length !== 1})\n        }};\n        " + code + "\n        return list;\n      });\n    ", { presets: ["es2015", "react", "stage-1"] }).code;
  };

  this._setTimeout = function () {
    for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    clearTimeout(_this2.timeoutID); //eslint-disable-line
    _this2.timeoutID = setTimeout.apply(null, args); //eslint-disable-line
  };

  this._executeCode = function () {
    var mountNode = _this2.refs.mount;

    try {
      (0, _reactDom.unmountComponentAtNode)(mountNode);
    } catch (e) {
      console.error(e); //eslint-disable-line
    }

    try {
      (function () {
        var scope = _this2.props.scope;

        var tempScope = [];
        Object.keys(scope).forEach(function (s) {
          return tempScope.push(scope[s]);
        });
        tempScope.push(mountNode);
        var compiledCode = _this2._compileCode();

        var Comp = function (_Component2) {
          _inherits(Comp, _Component2);

          function Comp() {
            var _ref2;

            var _temp2, _this3, _ret3;

            _classCallCheck(this, Comp);

            for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
              args[_key3] = arguments[_key3];
            }

            return _ret3 = (_temp2 = (_this3 = _possibleConstructorReturn(this, (_ref2 = Comp.__proto__ || Object.getPrototypeOf(Comp)).call.apply(_ref2, [this].concat(args))), _this3), _this3._createConsoleLine = function (_ref3) {
              var val = _ref3.val,
                  multipleArgs = _ref3.multipleArgs;
              return _react2.default.createElement(
                "span",
                { style: { marginRight: "20px" } },
                multipleArgs ? val.map(function (y) {
                  return _this3._createConsoleLine([y], false);
                }) : wrapMap["wrap" + getType(val[0])](val[0])
              );
            }, _temp2), _possibleConstructorReturn(_this3, _ret3);
          }

          _createClass(Comp, [{
            key: "render",
            value: function render() {
              var _this4 = this;

              return _react2.default.createElement(
                "div",
                { style: { padding: 15, fontFamily: "Consolas, Courier, monospace" } },
                eval(compiledCode).apply(null, tempScope).map(function (x, i) {
                  return (//eslint-disable-line
                    _react2.default.createElement(
                      "div",
                      {
                        key: i,
                        style: {
                          borderBottom: "1px solid #ccc",
                          padding: "4px 0"
                        } },
                      _this4._createConsoleLine(x)
                    )
                  );
                })
              );
            }
          }]);

          return Comp;
        }(_react.Component);

        (0, _reactDom.render)(_react2.default.createElement(Comp, null), mountNode);
      })();
    } catch (err) {
      _this2._setTimeout(function () {
        (0, _reactDom.render)(_react2.default.createElement(
          "div",
          { className: "playgroundError" },
          err.toString()
        ), mountNode);
      }, 500);
    }
  };

  this.componentDidMount = function () {
    _this2._executeCode();
  };

  this.componentDidUpdate = function (prevProps) {
    clearTimeout(_this2.timeoutID); //eslint-disable-line
    if (_this2.props.code !== prevProps.code) {
      _this2._executeCode();
    }
  };
};

exports.default = EsPreview;
