// react-docgen-displayname-handler/distribution/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDisplayNameHandler = createDisplayNameHandler;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _recast = require('recast');

var _recast2 = _interopRequireDefault(_recast);

var _reactDocgen = require('react-docgen');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getMemberValuePath = _reactDocgen.utils.getMemberValuePath;
var getNameOrValue = _reactDocgen.utils.getNameOrValue;
var types = _recast2.default.types.namedTypes;


var DEFAULT_NAME = 'UnknownComponent';

function getStaticDisplayName(path) {
  var displayName = null;
  var staticMember = getMemberValuePath(path, 'displayName');
  if (staticMember && types.Literal.check(staticMember.node)) {
    displayName = getNameOrValue(staticMember);
  }

  return displayName || null;
}

function getNodeIdentifier(path) {
  var displayName = null;
  if (types.FunctionExpression.check(path.node) || types.FunctionDeclaration.check(path.node) || types.ClassExpression.check(path.node) || types.ClassDeclaration.check(path.node)) {
    displayName = getNameOrValue(path.get('id'));
  }

  return displayName || null;
}

function getVariableIdentifier(path) {
  var displayName = null;
  var searchPath = path;

  while (searchPath !== null) {
    if (types.VariableDeclarator.check(searchPath.node)) {
      displayName = getNameOrValue(searchPath.get('id'));
      break;
    }
    searchPath = searchPath.parentPath;
  }

  return displayName || null;
}

function getNameFromFilePath() {
  var filePath = arguments.length <= 0 || arguments[0] === undefined ? '' : arguments[0];

  var displayName = null;

  var filename = _path2.default.basename(filePath, _path2.default.extname(filePath));
  if (filename === 'index') {
    var parts = _path2.default.dirname(filePath).split(_path2.default.sep);
    displayName = parts[parts.length - 1];
  } else {
    displayName = filename;
  }

  return displayName.charAt(0).toUpperCase().concat(displayName.slice(1)).replace(/-([a-z])/, function (_, match) {
    return match.toUpperCase();
  });
}

function createDisplayNameHandler(filePath) {
  return function displayNameHandler(documentation, path) {
    var displayName = [getStaticDisplayName, getNodeIdentifier, getVariableIdentifier].reduce(function (name, getDisplayName) {
      return name || getDisplayName(path);
    }, '');

    if (!displayName) {
      displayName = getNameFromFilePath(filePath);
    }

    documentation.set('displayName', displayName || DEFAULT_NAME);
  };
}

exports.default = createDisplayNameHandler('');
